src/library/autodj/autodjprocessor.cpp

```diff
@@ -723,155 +917,469 @@ bool AutoDJProcessor::removeTrackFromTopOfQueue(TrackPointer pTrack) {
+        introStart = getIntroStartPosition(pToDeck);
+    } else {
+        introStart = toDeckPosition;
+    }
 
-        // Guard against tracks shorter than the desired transition period.
-        // Use a sensible cap.
-        double nextTransitionTime = math_min3(desiredTransitionTime,
-                                              fromTrackDuration / 2,
-                                              toTrackDuration / 2);
-
-        double fromTrackStartFadeOut = fromTrackOutroEnd - nextTransitionTime;
+    double introEnd = getIntroEndPosition(pToDeck);
+    if (introEnd < introStart) {
+        // introEnd is invalid. Assume a zero length intro.
+        // The introStart is automatically placed by AnalyzerSilence, so use
+        // that as a fallback if the user has not placed introEnd.
+        introEnd = introStart;
+    }
 
-        if (fromTrackDuration > 0.0) {
-            pFromDeck->posThreshold = fromTrackStartFadeOut / fromTrackDuration;
-            pFromDeck->fadeDuration = nextTransitionTime / fromTrackDuration;
+    double introLength = introEnd - introStart;
+
+    switch (m_transitionMode) {
+    case TransitionMode::FullIntroOutro:
+        // Use the outro or intro length for the transition time, whichever is
+        // shorter. Let the full outro and intro play; do not cut off any part
+        // of either.
+        //
+        // In the diagrams below,
+        // - is part of a track outside the outro/intro,
+        // o is part of the outro
+        // i is part of the intro
+        // | marks the boundaries of the transition
+        //
+        // When outro > intro:
+        // ------ooo|ooo|
+        //          |iii|------
+        //
+        // When outro < intro:
+        // ------|ooo|
+        //       |iii|iii-----
+        //
+        // If only the outro or intro length is marked but not both, use the one
+        // that is marked for the transition time. If neither is marked, fall
+        // back to the transition time from the spinbox.
+        if (outroLength > 0 && introLength > 0) {
+            if (outroLength > introLength) {
+                pFromDeck->fadeBeginPos = outroEnd - introLength;
+                pFromDeck->fadeEndPos = outroEnd;
+                pToDeck->startPos = introStart;
+            } else {
+                pFromDeck->fadeBeginPos = outroStart;
+                pFromDeck->fadeEndPos = outroEnd;
+                pToDeck->startPos = introStart;
+            }
+        } else if (outroLength > 0 && introLength <= 0) {
+            if (outroLength + introStart < pToDeck->fadeBeginPos) {
+                pFromDeck->fadeBeginPos = outroStart;
+            } else {
+                // This happens if the toDeck track has no intro set and the
+                // outro of the fromDeck track is longer than the whole toDeck
+                // track
+                outroLength = pToDeck->fadeBeginPos - introStart;
+                VERIFY_OR_DEBUG_ASSERT(outroLength > 0) {
```

Comment:
This debug assertion is triggered inside a lot of the AutoDJ processor tests: https://travis-ci.org/github/mixxxdj/mixxx/jobs/705774858#L2692
You notice it when compiling with #2911. 

@daschuer You introduced this assertion in 010fd5cf847303549086eeb5dac9258eb9e24462 and know the AutoDJ code better than I do. Can you have a look?

Reply:
Or maybe @Be-ing check this?